### PR TITLE
Initial data logging implementation.

### PR DESCRIPTION
**Pull request by microbit-matt-hillsdon**
This can now also be reviewed via the UI PR: https://github.com/microbit-foundation/python-editor-next/pull/922

### CODE REVIEW

File: src/board/data-logging.ts [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+  MICROBIT_HAL_LOG_TIMESTAMP_HOURS,
+  MICROBIT_HAL_LOG_TIMESTAMP_MILLISECONDS,
+  MICROBIT_HAL_LOG_TIMESTAMP_MINUTES,
+  MICROBIT_HAL_LOG_TIMESTAMP_NONE,
+  MICROBIT_HAL_LOG_TIMESTAMP_SECONDS,
+} from "./constants";
+import { State } from "./state";
+
+// TODO: Check this value
+const maxSize = 124 * 1024;
```
Comment (microbit-matt-hillsdon) [inline src/board/data-logging.ts]: Need to figure out the max size for the log.

File: src/board/data-logging.ts [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    ? entry.headings.reduce((prev, curr) => prev + curr.length, 0)
+    : 0;
+  const data = entry.data
+    ? entry.data.reduce((prev, curr) => prev + curr.length, 0)
+    : 0;
+  return headings + data;
+}
+
+function timestampHeading(timestamp: number): string {
+  let unit: string;
```
Comment (microbit-matt-hillsdon) [inline src/board/data-logging.ts]: I extracted this so I could use "return" and then... failed to.

File: src/board/state.ts [hunk 3]
```diff
@@ -97,3 +97,38 @@ export class EnumSensor extends Sensor {
...
+    return {
+      type: "log",
+      id: this.id,
+      value: this.value,
+      period: this.period,
+      serial: this.serial,
+      delete: this.delete,
+    };
+  }
+}
```
Comment (microbit-matt-hillsdon) [inline src/board/state.ts]: Not used.

File: src/microbithal_js.c [hunk 4]
```diff
@@ -386,18 +386,21 @@ const uint8_t *microbit_hal_get_font_data(char c) {
 }
 
 void microbit_hal_log_delete(bool full_erase) {
+    mp_js_hal_log_delete(full_erase);
     /*
     uBit.log.clear(full_erase);
     */
```
Comment (microbit-matt-hillsdon) [inline src/microbithal_js.c]: I'll remove these.

File: src/board/data-logging.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    if (!this.row) {
+      throw noRowError();
+    }
+    if (this.timestamp !== this.timestampOnLastEndRow) {
+      // New timestamp column required. Put it first if there's been no output.
+      if (this.size === 0) {
+        this.headings = [timestampToHeading(this.timestamp), ...this.headings];
+        this.row = ["", ...this.row];
+      } else {
+        this.logData(timestampToHeading(this.timestamp), "");
```
Comment (microbit-robert) [inline src/board/data-logging.ts:64]: What's the purpose of this line?

**Comment by microbit-robert** [comment]
I think the default timestamp should be milliseconds, but the example logs out seconds without changing the timestamp?

**Comment by microbit-matt-hillsdon** [comment]
> I think the default timestamp should be milliseconds, but the example logs out seconds without changing the timestamp?

This is correct (due to set_labels call) but the docs are wrong. Raised https://github.com/microbit-foundation/python-editor-next/issues/920

File: src/board/data-logging.ts (hunk 5 above)
Comment (microbit-matt-hillsdon) [reply src/board/data-logging.ts:64]: It allows the timestamps to change after the first row has been written.

Comment (microbit-matt-hillsdon) [reply src/board/data-logging.ts:64]: See https://github.com/lancaster-university/codal-microbit-v2/blob/master/source/MicroBitLog.cpp#L554

**Review by microbit-robert — APPROVED** [review]
LGTM.

**Pull request merged**